### PR TITLE
mat: Expose ReuseAs method to help resizing

### LINKED ISSUE #963: mat: Helper routine(s) for resizing zero matrices
We have a pattern that Zero matrices can be resized. This is both receivers of methods, but also as inputs to functions, for example in `stat` and `optimize`. It would be nice to implement methods necessary to make this pattern easy to implement. This would encourage outside packages to follow the same pattern.

### PR DESCRIPTION
**Pull request by btracey**
Fixes #963

Please take a look.

### CODE REVIEW

**Comment by btracey** [comment]
I started the implementation, but wanted to discuss the singular case (documentation and behavior) before doing the work on all of the matrix types

1) I think ReuseAs should be Zeroed. This is going to be far safer in the general case, and I fear not having it be that way will eventually be a source of bugs. If we need it in the future we could add a ReuseAsNonzerod which at least is explicit about the behavior

2) If we're agreed on 1), this is going to get confusing with the unexported code. I propose to make the name shift reuseAs -> reuseAsNZ in this PR (and so all uses will be caught with a compile time error)

File: mat/dense.go [hunk 1]
```diff
@@ -67,6 +67,27 @@ func NewDense(r, c int, data []float64) *Dense {
...
+// Reset should be used.
+func (m *Dense) ReuseAs(r, c int) {
+	if r <= 0 || c <= 0 {
+		if r == 0 || c == 0 {
+			panic(ErrZeroLength)
+		}
+		panic(ErrNegativeDimension)
+	}
+	if !m.IsZero() {
+		panic("mat: ReuseAs called on non-zero matrix")
```
Comment (btracey) [inline mat/dense.go]: This should be made one of the Error types

**Comment by kortschak** [comment]
1. Agreed.
2. I don't understand the basis for this possible confusion.

File: mat/dense.go (hunk 1 above)
Comment (kortschak) [reply mat/dense.go]: Yes, with the text "mat: reuse non-zero matrix".

**Comment by btracey** [comment]
The basis is that `ReuseAs` would zero while `reuseAs` would not (and instead `reuseAsZeroed` gets the same behavior)

**Comment by kortschak** [comment]
Ah, yes. Thanks. Agreed.

**Comment by btracey** [comment]
PTAL 

This implements the method for Dense

File: mat/symmetric.go [hunk 2]
```diff
@@ -176,6 +197,11 @@ func (s *SymDense) reuseAs(n int) {
 	}
 }
 
+func (s *SymDense) reuseAsZeroed(n int) {
```
Comment (kortschak) [inline mat/symmetric.go]: This should be fully spelled out to avoid double work on the backing slice.

File: mat/errors.go [hunk 3]
```diff
@@ -114,7 +114,9 @@ type Error struct{ string }
 func (err Error) Error() string { return err.string }
 
 var (
+	ErrNegativeDimension   = Error{"matrix: negative dimension"}
```
Comment (kortschak) [inline mat/errors.go]: Should we fix the prefixes here or in another PR?

Comment (btracey) [reply mat/errors.go]: Filed https://github.com/gonum/gonum/issues/1084

**Comment by btracey** [comment]
`reuseAs` also allows the matrix to be size rxc. In internal code, this allows  `m.reuseAs(r, c)` without the extra check to the size of the matrix. Do we want to replicate this behavior?

```
func (r *Foo) Bar(dst *mat.Dense){
    if dst.IsZero() {
         dst.ReuseAs(r, c)
    }else {
        if r2, c2 := dst.Dims(); r != r2 || c != c2 {
              panic("receiver shape mismatch")
        }
    }
    // do stuff
}
```

The alternative is just

```
func (f *Foo) Bar(dst *mat.Dense){
     dst.ReuseAs(r, c)
     // do stuff
}
```

The former is clearly more code and so slightly harder to use. The upside is it makes the `ReuseAs` method easier to understand, and will make panic traces easier to parse. 

Thoughts?

File: mat/symmetric.go [hunk 4]
```diff
@@ -176,6 +197,11 @@ func (s *SymDense) reuseAs(n int) {
...
+	s.reuseAs(n)
```
Comment (btracey) [inline mat/symmetric.go]: This needs to change to reuseAsNonZeroed

**Comment by kortschak** [comment]
In the first examples, should that be `dst.reuseAs(r, c)` (unexported)?

File: mat/symmetric.go (hunk 4 above)
Comment (kortschak) [reply mat/symmetric.go]: I'd rather we don't do it that way but just spell out this completely.

Comment (kortschak) [reply mat/symmetric.go]: Actually, `useZeroed` is incorrect for the general case. So leave this as is and we'll fix `useZeroed` later. The issue is that the `zero` function is unaware of striding and so will clobber other data outside the matrix if they share a backing slice.

Comment (kortschak) [reply mat/symmetric.go]: Updating this, it's not fixable in itself, so this should be updated. We need some additional documentation as a warning on the `ReuseAs` methods. They are a fairly substantial footgun that cannot have a safety.

File: mat/dense.go [hunk 5]
```diff
@@ -67,11 +67,33 @@ func NewDense(r, c int, data []float64) *Dense {
...
-// reuseAs resizes an empty matrix to a r×c matrix,
-// or checks that a non-empty matrix is r×c.
+// ReuseAs changes the receiver if it IsZero() to be of size r×c.
+//
+// ReuseAs re-uses the backing data slice if it has sufficient capacity,
+// otherwise a new slice is allocated. The data is then zeroed.
+//
+// ReuseAs panics if the receiver is not zero-sized, and panics if
+// the input sizes are less than one. To zero a matrix for re-use,
+// Reset should be used.
```
Comment (kortschak) [inline mat/dense.go]:
```
//
// ReuseAs is not safe to use if the receiver shares backing data with
// another Matrix.
```

File: mat/symmetric.go [hunk 6]
```diff
@@ -134,6 +134,27 @@ func (s *SymDense) Reset() {
...
 }
 
+// ReuseAsSym changes the receiver if it IsZero() to be of size n×n.
+//
+// ReuseAsSym re-uses the backing data slice if it has sufficient capacity,
+// otherwise a new slice is allocated. The data is then zeroed.
+//
+// ReuseAsSym panics if the receiver is not zero-sized, and panics if
+// the input size is less than one. To zero a matrix for re-use,
+// Reset should be used.
```
Comment (kortschak) [inline mat/symmetric.go]:
```
//
// ReuseAsSym is not safe to use if the receiver shares backing data with
// another Matrix.
```

**Comment by btracey** [comment]
No, I meant the exported function. The question is how do outside packages interact with this function. The goal is that that if I'm implemented `mypgk.Foo`, and want to resize a destination matrix, what is the pattern? (For example, in `stat`, how do we want the checking code to look?)

File: mat/symmetric.go (hunk 4 above)
Comment (btracey) [reply mat/symmetric.go]: Sorry, that was a note to myself that `SymDense.reuseAs` needs to also be changed to `SymDense.reuseAsNonZeroed`

File: mat/dense.go (hunk 5 above)
Comment (btracey) [reply mat/dense.go]: Isn't this obvious from the behavior? The same is true with `Reset`, `Augment`, `Copy`, `CloneFrom`, etc. Any operation that can change the backing slice is not safe to use when sharing backing data.

Comment (kortschak) [reply mat/dense.go]: It is also true for those. I think that given that people mess up `append` and this is more complex, it's worth noting. Probably in the package doc as well that these are possible. The issue is not about shared data though, it's about potentially clearing data that's not obviously shared.

**Comment by kortschak** [comment]
Externally (outside gonum), it should look like the bottom one. Inside gonum, probably the upper one to make the panics clearer.

File: mat/dense.go (hunk 5 above)
Comment (btracey) [reply mat/dense.go]: I mean, it is about shared data, in the sense that if you don't share data there are no problems. There also aren't any problems if you aren't tricky about it (splitting rows and columns of matrices, which is actually reasonably hard to do with the package).

I don't think this function itself presents problems. The dangerous one is `Reset`, which enables `ReuseAs` and 85% of the potential for misuse of shared data (i.e. reshaping receivers and destination matrices).

Comment (kortschak) [reply mat/dense.go]: The second paragraph is the issue, yes. Perhaps the warning should be on reset. There the warning would be that by resetting matrix guards against leaky sharing are lost.

File: mat/dense.go (hunk 1 above)
Comment (btracey) [reply mat/dense.go]: Done.

File: mat/symmetric.go (hunk 2 above)
Comment (btracey) [reply mat/symmetric.go]: Done.

**Comment by btracey** [comment]
Okay. Done.

1) Added some extra warnings to Reset
2) Exposed ReuseAs on the types where we internally check for sizes (Dense, Tri, Sym, Vec). Leaving Diag and the bandeds for later.
3) Added some helper types for making these kinds of consistent interfaces to check in the future. This was made as simple as possible, and not intended to be comprehensive. In particular I did not add denseMatrix to VecDense because VecDense does not implement DiagView or Tracer. I figured it was outside the scope of this PR to talk about that here.

File: mat/symmetric.go (hunk 6 above)
Comment (btracey) [reply mat/symmetric.go]: Added warnings to Reset instead.

**Comment by kortschak** [comment]
Fixes #1085.

**Pull request merged**